wnfs/src/private/hamt/diff/key_value.rs

```diff
@@ -66,105 +68,200 @@ type EitherPairOrNode<'a, K, V, H> = Option<Either<&'a Pair<K, V>, &'a Rc<Node<K
-                }
-                _ => unreachable!("Node change type is Modify but nodes not found or not pairs."),
-            },
+            (true, true) => {
+                // Main and other have a value. They may be the same or different so we check.
+                let main_index = main_node.get_value_index(index);
+                let main_pointer = mem::take(main_node.pointers.get_mut(main_index).unwrap());
+
+                let other_index = other_node.get_value_index(index);
+                let other_pointer = mem::take(other_node.pointers.get_mut(other_index).unwrap());
+
+                changes.extend(
+                    generate_modify_changes(main_pointer, other_pointer, depth, store).await?,
+                );
+            }
+            (false, false) => { /* No change */ }
         }
     }
 
-    Ok(kv_changes)
+    Ok(changes)
 }
 
-async fn generate_add_or_remove_changes<'a, K, V, H, B>(
-    node: EitherPairOrNode<'a, K, V, H>,
+async fn generate_add_or_remove_changes<K, V, H>(
+    node_pointer: &Pointer<K, V, H>,
     r#type: ChangeType,
-    store: &B,
+    store: &mut impl BlockStore,
 ) -> Result<Vec<KeyValueChange<K, V>>>
 where
-    B: BlockStore,
-    K: DeserializeOwned + Clone,
-    V: DeserializeOwned + Clone,
+    K: DeserializeOwned + Clone + Eq + Hash + AsRef<[u8]>,
+    V: DeserializeOwned + Clone + Eq,
     H: Hasher + Clone + 'static,
 {
-    match node {
-        Some(Left(Pair { key, value })) => Ok(vec![KeyValueChange {
-            r#type,
-            key: key.clone(),
-            main_value: if r#type == ChangeType::Add {
-                Some(value.clone())
-            } else {
-                None
-            },
-            other_value: if r#type == ChangeType::Remove {
-                Some(value.clone())
-            } else {
-                None
-            },
-        }]),
-        Some(Right(node)) => {
-            node.flat_map(
-                &|Pair { key, value }| {
-                    Ok(KeyValueChange {
-                        r#type,
+    match node_pointer {
+        Pointer::Values(values) => Ok(values
+            .iter()
+            .map(|Pair { key, value }| KeyValueChange {
+                r#type,
+                key: key.clone(),
+                value1: Some(value.clone()),
+                value2: None,
+            })
+            .collect()),
+        Pointer::Link(link) => {
+            let node = link.resolve_value(store).await?;
+            node.as_ref()
+                .flat_map(
+                    &|Pair { key, value }| {
+                        Ok(KeyValueChange {
+                            r#type,
+                            key: key.clone(),
+                            value1: Some(value.clone()),
+                            value2: None,
+                        })
+                    },
+                    store,
+                )
+                .await
+        }
+    }
+}
+
+async fn generate_modify_changes<K, V, H>(
```

Comment:
I didn't think the function name was 100% fitting, so here's a suggestion:

```suggestion
async fn pointers_kv_diff<K, V, H>(
```

That said, this is an internal helper method, so not super important.